components/omega/doc/design/Metadata.md

```diff
@@ -0,0 +1,320 @@
+the code, simulation, file metadata, but also for variables if additional
+metadata is desired. This will be an aliased function based on
+the data type of the metadata.
+
+    ierr = MetadataAdd(varName, name, value);
+
+where varName is the name of the already defined variable, and
+the other arguments are the usual name/value pair for the metadata
+to be added. This function will return 0 if successful and an error
+if either the varName has not been defined or if a metadata pair with
+that name already exists (or maybe we want to allow overwrite?).
+
+For symmetry, we will supply a remove function, though the
+use case is likely rare.
+
+    ierr = MetadataRemove(varName, name); 
+
+This doesn't need a type, though it might make it easier
+to search.
+
+#### 4.2.3 Retrieve metadata
+
+The most common use case will be retrieving metadata. For a single
+metadata entry, there will be an explicit get function:
+
+    myValue = MetadataGet(varName, name);
+
+where varName is the field defined (or generic name for global
+metadata) and name is the name associated with the metadata
+to be retrieved.
+
+During I/O stages, we will need a capability for retrieving
+all metadata for a defined field or global category. We will
+supply a function to retrieve the map associated with a given
+data type. The `std::map` iterators and functionality can then
+be used to extract all of the metadata for a given type. For
+example, to retrieve all R4 metadata from a defined field:
+
+   std::map<std::string, R4> r4meta = MetadataGetAll(varName);
+   std::string metaName;
+   R4 metaVal
+
+   // use std::map functionality to loop through all metadata
+   for (auto it=r4meta.begin(); it != r4.meta.end(), ++it){
+      metaName = it->first;  // retrieve name part of meta pair
+      metaVal  = it->second; // retrieve value part of meta pair
+
+      // do whatever needed with metadata
+   }
+
+#### 4.2.4 Create dimension
+
+For the dimension class, there will be a constructor that will
+both construct the instance and add it to the list of defined
+dimensions. It will generate an error if the dimension is already
+defined.
+
+    MetadataDim(std::string name, // name of dimension
+                I4 length         // length of dimension
+                );
+
+A destructor will be supplied as well.
+
+#### 4.2.5 Retrieve dim length
+
+The only other function for a dimension will be a retrieval for
+the dimension length:
+
+    I4 myLength = myDim("name");
+
+#### 4.2.6 Existence inquiry
+
+For both metadata and dimensions, a function will be provided
+to inquire whether metadata or dimensions with a given name
+have already been defined:
+
+    bool isMetadataDefined(const std::string name);
+    bool isDimDefined     (const std::string name); 
+
+## 5 Verification and Testing
+
+Requirement 2.3 will be enforced as needed when defining available
+fields so will not be tested as part of unit testing.
```

Comment:
It could still be a good idea to run a cf-checker either in Omega or in polaris.  It seems like it might be easier to do it on the polaris side (so not as a unit test).  But it seems pretty likely that some CF non-compliance will emerged by mistake in the development process so we do want some way to catch this.

Reply:
added text for this (though probably will be a polaris test)